Config entry and device for Coolmaster integration

OnFreund opened the pull request:
## Breaking Change:

Coolmaster configuration is now taken from config entry rather than `configuration.yaml`.
If you're using the CoolMasterNet integration, you'll need to remove the configuration from `configuration.yaml`, and add the integration through the Integrations UI. Click on the `+` sign at the bottom right, select 'CoolMasterNet`, and configure your host and supported modes.

## Description:
This PR creates devices for individual Coolmaster units. This allows them, for example, to be placed in areas.
In order to support devices, a config flow is also implemented, and the configuration is in a config entry.

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#10947

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [ ] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

OnFreund committed:
Config entry and device for Coolmaster integration

OnFreund commented:
When running this locally, I couldn't get the config flow to show the strings from strings.json. Not sure how to fix that, or if it's like that in dev (also tried a new integration using scaffold and the result was the same)

OnFreund committed:
Lint/isort/flake/etc...

OnFreund committed:
Black formatting

balloob commented:
Run `script/translations_develop` to get the translation files updated for the UI. No need to commit them.

balloob commented on homeassistant/components/coolmaster/__init__.py:
```diff
@@ -1 +1,20 @@
+
+async def async_unload_entry(hass, entry):
+    """Unload a Coolmaster config entry."""
+    hass.async_add_job(hass.config_entries.async_forward_entry_unload(entry, "climate"))
```
await unloading.

balloob commented on homeassistant/components/coolmaster/climate.py:
```diff
@@ -69,18 +47,22 @@ def _build_entity(device, supported_modes):
-def setup_platform(hass, config, add_entities, discovery_info=None):
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
     """Set up the CoolMasterNet climate platform."""
+    return True
```
There is no return value. You can just remove the whole method.

balloob commented on homeassistant/components/coolmaster/climate.py:
```diff
@@ -118,6 +100,16 @@ def update(self):
+        return {
+            "identifiers": {(DOMAIN, self.unique_id)},
+            "name": self.name,
+            "manufacturer": "CoolAutomation",  # TODO: take from `line`
```
todos

OnFreund replied:
What's the preferred way of documenting those? It required changes in the 3rd party lib, so not likely to happen soon, but still I believe it deserves a mention for future maintenance.

balloob replied:
Remove them. Person who will update it in the future can figure it out .

balloob commented on homeassistant/components/coolmaster/config_flow.py:
```diff
@@ -0,0 +1,44 @@
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        if user_input is not None:
+            return self._async_get_entry(user_input)
```
You need to validate that the input works.

OnFreund replied:
I added validation, but when showing an error the submit button is then outside the view. Not sure how to resolve that.

balloob replied:
screenshot ?

OnFreund replied:
Before:
<img width="499" alt="before" src="https://user-images.githubusercontent.com/2159693/67260915-b3cfcf80-f4a6-11e9-837e-2d1e0fcaa193.png">
After:
<img width="517" alt="after" src="https://user-images.githubusercontent.com/2159693/67260919-b7635680-f4a6-11e9-9397-4f1779dc4e9d.png">

balloob commented on homeassistant/components/coolmaster/config_flow.py:
```diff
@@ -0,0 +1,44 @@
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+}
+
+DATA_SCHEMA = vol.Schema({**HOST_SCHEMA, **MODES_SCHEMA})
```
Just inline HOST_SCHEMA in here ?

OnFreund replied:
That's how I had it originally, but it didn't look good. Python is not a language I'm proficient with, so happy to rewrite this in a more idiomatic way if one exists.

balloob reviewed (changes requested):
Looks great, few minor comments ! Don't forget to also add a test for the config flow. You can use `python3 -m script.scaffold config_flow` to generate a test foundation for your config flow.

OnFreund committed:
Code review fixes

OnFreund commented:
Thanks @balloob. I addressed most of your comments (will also add test, and waiting for your input on TODO).
One thing I noticed is that when adding the integration, all entities are found, but only a single device gets added (tried several times, and it was a different device each time). After restarting HA, the devices for all entities are found.

balloob commented:
The devices will show up as soon as Home Assistant is made aware of them. Do you call async_add_entitites right away when integration is set up ?

OnFreund commented:
I'm not calling it explicitly. I'm calling `async_create_entry`, and then `async_setup_entry` is calling it. However, the entities are being added, it's just the devices that aren't, or more accurately, just one of them is added, and it's a different one each time.

OnFreund committed:
Config flow tests for coolmaster

OnFreund committed:
Add pycoolmaster requirement to test

OnFreund commented:
Not sure why pylint is complaining that `DOMAIN` is not used in `config_flow.py` - it clearly is used.

balloob commented:
I think that's a bug in PyLint. Ignore it.

For port, now it's a slider and that is pretty dumb. Any reason why it should be configurable to begin with, wouldn't the port always be the same ?

OnFreund commented:
I was thinking about the slider too. Technically the port can be changed, don't know how common that is.
However, I'm assuming port selection is going to be common with config flow integrations. Maybe we need a different UI for it? Maybe an up/down box?

balloob commented:
Can't we just drop it and use the default port ?

OnFreund commented:
I don't know how many people use this integration, and whether anyone of them is using a custom port. We could release without it, and if anyone complains, quickly add this to a 0.X.1 release.
As mentioned above, though, I'm assuming the issue of port selection is going to come up often with other integrations.

balloob commented:
Port selection doesn't come up that often because most services do not allow changing ports.

OnFreund committed:
Remove port selection from Coolmaster config flow

OnFreund commented:
OK, I removed it, which also fixes the `submit` button going out of view. Keep in mind that other integrations do have port selection (for example, AdGuard & MQTT)

balloob commented on homeassistant/components/coolmaster/config_flow.py:
```diff
@@ -0,0 +1,63 @@
+
+MODES_SCHEMA = {vol.Required(mode, default=True): bool for mode in AVAILABLE_MODES}
+
+DATA_SCHEMA = vol.Schema({**{vol.Required(CONF_HOST): str}, **MODES_SCHEMA})
```
```suggestion
DATA_SCHEMA = vol.Schema({vol.Required(CONF_HOST): str, **MODES_SCHEMA})
```

balloob committed:
Update config_flow.py

OnFreund committed:
More idoimatic hash concat

MartinHjelmare commented on homeassistant/components/coolmaster/__init__.py:
```diff
@@ -1 +1,22 @@
+
+async def async_setup_entry(hass, entry):
+    """Set up Coolmaster from a config entry."""
+    hass.async_add_job(hass.config_entries.async_forward_entry_setup(entry, "climate"))
```
`hass.async_add_job` is legacy for general use. Please use `hass.async_create_task` instead to schedule a coroutine task on the event loop.

MartinHjelmare commented on homeassistant/components/coolmaster/__init__.py:
```diff
@@ -1 +1,22 @@
+
+async def async_unload_entry(hass, entry):
+    """Unload a Coolmaster config entry."""
+    await hass.async_add_job(
```
We don't need to create a task here. Just await the unloading directly. Also return the awaited return value since it's a boolean indicating success.

MartinHjelmare commented on homeassistant/components/coolmaster/config_flow.py:
```diff
@@ -0,0 +1,64 @@
+    """Validate that we can connect to the Coolmaster instance."""
+    cool = CoolMasterNet(host, port=DEFAULT_PORT)
+    devices = await hass.async_add_executor_job(cool.devices)
+    return len(devices) > 0
```
```py
return bool(devices)
```
or just
```py
return devices
```

MartinHjelmare commented on homeassistant/components/coolmaster/climate.py:
```diff
@@ -50,17 +39,6 @@
 FAN_MODES = ["low", "med", "high", "auto"]
 
-CONF_SUPPORTED_MODES = "supported_modes"
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
This is a breaking change.

MartinHjelmare commented on tests/components/coolmaster/test_config_flow.py:
```diff
@@ -0,0 +1,104 @@
+    assert result["errors"] is None
+
+    with patch(
+        "homeassistant.components.coolmaster.config_flow.validate_connection",
```
We should patch the coolmaster library not our code.

MartinHjelmare commented on tests/components/coolmaster/test_config_flow.py:
```diff
@@ -0,0 +1,104 @@
+    )
+
+    with patch(
+        "homeassistant.components.coolmaster.config_flow.validate_connection",
```
See above.

MartinHjelmare commented on tests/components/coolmaster/test_config_flow.py:
```diff
@@ -0,0 +1,104 @@
+    )
+
+    with patch(
+        "homeassistant.components.coolmaster.config_flow.validate_connection",
```
See above.

MartinHjelmare commented on tests/components/coolmaster/test_config_flow.py:
```diff
@@ -0,0 +1,104 @@
+    )
+
+    with patch(
+        "homeassistant.components.coolmaster.config_flow.validate_connection",
```
See above.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR.

MartinHjelmare commented:
Please update the breaking change paragraph with what the user needs to do to cope with the breaking change. Ie, remove the config yaml entry and go to integrations panel in the GUI to set up the integration.

OnFreund commented:
I updated the breaking change paragraph, and created PR https://github.com/home-assistant/home-assistant/pull/28163 to address your comment
